osu.Game/Online/OnlineViewContainer.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Allocation;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Online.API;
+using osu.Game.Online.Placeholders;
+
+namespace osu.Game.Online
+{
+    /// <summary>
+    /// A <see cref="Container"/> for dislaying online content who require a local user to be logged in.
+    /// Shows its children only when the local user is logged in and supports displaying a placeholder if not.
+    /// </summary>
+    public class OnlineViewContainer : Container, IOnlineComponent
+    {
+        private readonly Container placeholderContainer;
+        private readonly Placeholder placeholder;
+        private readonly LoadingAnimation loading;
+
+        private const int transform_time = 300;
+
+        protected Container View { get; }
+        protected override Container<Drawable> Content => View;
+
+        [Resolved]
+        protected IAPIProvider API { get; private set; }
+
+        public OnlineViewContainer(string placeholderMessage)
+        {
+            InternalChildren = new Drawable[]
+            {
+                View = new Container
+                {
+                    RelativeSizeAxes = Axes.Both,
+                },
+                placeholderContainer = new Container
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Alpha = 0,
+                    Child = placeholder = new LoginPlaceholder(placeholderMessage)
+                },
+                loading = new LoadingAnimation
+                {
+                    Alpha = 0,
+                }
+            };
+        }
+
+        public virtual void APIStateChanged(IAPIProvider api, APIState state) => UpdatePlaceholderVisibility(state);
+
+        /// <summary>
+        /// Updates the visibility of the placeholder for this <see cref="OnlineViewContainer"/> accordingly to the passed <see cref="APIState"/>.
+        /// <list type="bullet">
+        /// <item> Passing <see cref="APIState.Online"/> will cause the online-only view to display.</item>
+        /// <item> Passing <see cref="APIState.Offline"/> will cause the placeholder to display.</item>
+        /// <item> Passing <see cref="APIState.Connecting"/> or <see cref="APIState.Failing"/> will cause the loading animation to display.</item>
+        /// </list>
+        /// </summary>
+        /// <param name="status">The new API state</param>
+        protected void UpdatePlaceholderVisibility(APIState status)
```

Comment:
Is there a reason this is protected? I'm not sure if it's fine for a deriver to be able to manipulate the placeholder visibility.